wallets/dcr/sync.go

```diff
@@ -0,0 +1,484 @@
+			return wallet.syncData.headersRescanProgress.GeneralSyncProgress
+		case CFiltersFetchSyncStage:
+			return wallet.syncData.cfiltersFetchProgress.GeneralSyncProgress
+		}
+	}
+
+	return nil
+}
+
+func (wallet *Wallet) ConnectedPeers() int32 {
+	wallet.syncData.mu.RLock()
+	defer wallet.syncData.mu.RUnlock()
+	return wallet.syncData.connectedPeers
+}
+
+func (wallet *Wallet) PeerInfoRaw() ([]PeerInfo, error) {
+	if !wallet.IsConnectedToDecredNetwork() {
+		return nil, errors.New(ErrNotConnected)
+	}
+
+	syncer := wallet.syncData.syncer
+
+	infos := make([]PeerInfo, 0, len(syncer.GetRemotePeers()))
+	for _, rp := range syncer.GetRemotePeers() {
+		info := PeerInfo{
+			ID:             int32(rp.ID()),
+			Addr:           rp.RemoteAddr().String(),
+			AddrLocal:      rp.LocalAddr().String(),
+			Services:       fmt.Sprintf("%08d", uint64(rp.Services())),
+			Version:        rp.Pver(),
+			SubVer:         rp.UA(),
+			StartingHeight: int64(rp.InitialHeight()),
+			BanScore:       int32(rp.BanScore()),
+		}
+
+		infos = append(infos, info)
+	}
+
+	sort.Slice(infos, func(i, j int) bool {
+		return infos[i].ID < infos[j].ID
+	})
+
+	return infos, nil
+}
+
+func (wallet *Wallet) PeerInfo() (string, error) {
+	infos, err := wallet.PeerInfoRaw()
+	if err != nil {
+		return "", err
+	}
+
+	result, _ := json.Marshal(infos)
+	return string(result), nil
+}
+
+func (wallet *Wallet) GetBestBlock() *BlockInfo {
+	var bestBlock int32 = -1
+	var blockInfo *BlockInfo
+	if !wallet.WalletOpened() {
+		return nil
+	}
+
+	walletBestBLock := wallet.getBestBlock()
+	if walletBestBLock > bestBlock || bestBlock == -1 {
+		bestBlock = walletBestBLock
+		blockInfo = &BlockInfo{Height: bestBlock, Timestamp: wallet.GetBestBlockTimeStamp()}
+	}
+
+	return blockInfo
+}
+
+func (wallet *Wallet) GetLowestBlock() *BlockInfo {
+	var lowestBlock int32 = -1
+	var blockInfo *BlockInfo
+	if !wallet.WalletOpened() {
+		return nil
+	}
+	walletBestBLock := wallet.getBestBlock()
+	if walletBestBLock < lowestBlock || lowestBlock == -1 {
+		lowestBlock = walletBestBLock
+		blockInfo = &BlockInfo{Height: lowestBlock, Timestamp: wallet.GetBestBlockTimeStamp()}
+	}
+
+	return blockInfo
+}
+
+func (wallet *Wallet) getBestBlock() int32 {
+	if wallet.Internal() == nil {
+		// This method is sometimes called after a wallet is deleted and causes crash.
+		log.Error("Attempting to read best block height without a loaded wallet.")
+		return 0
+	}
+
+	_, height := wallet.Internal().MainChainTip(wallet.ShutdownContext())
+	return height
+}
+
+func (wallet *Wallet) GetBestBlockTimeStamp() int64 {
+	if wallet.Internal() == nil {
+		// This method is sometimes called after a wallet is deleted and causes crash.
+		log.Error("Attempting to read best block timestamp without a loaded wallet.")
+		return 0
+	}
+
+	ctx := wallet.ShutdownContext()
+	_, height := wallet.Internal().MainChainTip(ctx)
+	identifier := w.NewBlockIdentifierFromHeight(height)
+	info, err := wallet.Internal().BlockInfo(ctx, identifier)
```

Comment:
Here you've made 3 calls to `wallet.Internal()`

Reply:
same comments as above